packages/trade-widget/src/sagas/blockchain_saga.ts

```diff
@@ -0,0 +1,99 @@
+import { ZeroEx } from '0x.js';
+import { BigNumber } from '@0xproject/utils';
+import { Web3Wrapper } from '@0xproject/web3-wrapper';
+import * as _ from 'lodash';
+import { Store } from 'redux';
+
+import { Dispatcher } from '../redux/dispatcher';
+import { State } from '../redux/reducer';
+import { AccountTokenBalances, AccountWeiBalances } from '../types';
+
+const POLLING_INTERVAL = 2000;
+const BALANCE_POLLING_INTERVAL = 5000;
+
+/**
+ * Watch for user account changes and network changes
+ * which are outside of our control. I.e injected web3 providers, balance changes
+ */
+export class BlockchainSaga {
+    private _store: Store<State>;
+    private _dispatch: Dispatcher;
+    private _web3Wrapper: Web3Wrapper;
+    private _zeroEx: ZeroEx;
+    private _prevState: State;
+    constructor(dispatch: Dispatcher, store: Store<State>, web3Wrapper: Web3Wrapper, zeroEx: ZeroEx) {
+        this._dispatch = dispatch;
+        this._web3Wrapper = web3Wrapper;
+        this._store = store;
+        this._zeroEx = zeroEx;
+        this._startPolling().catch(console.log);
+        store.subscribe(this._handleStateChange.bind(this));
+        this._handleStateChange();
+    }
+    // The intention here is to stop any long running processes
+    // if they are not relevant anymore. I.e watching for balance changes
+    // on an old network.
+    private _handleStateChange() {
+        const nextState = this._store.getState();
+        const lastAction = nextState.lastAction;
+        if (nextState !== this._prevState) {
+            this._prevState = nextState;
+        }
+    }
+    private async _startPolling() {
+        await this._checkBlockchainAsync();
+        await this._checkBalancesAsync();
+        setInterval(this._checkBlockchainAsync.bind(this), POLLING_INTERVAL);
+        setInterval(this._checkBalancesAsync.bind(this), BALANCE_POLLING_INTERVAL);
+    }
+    private async _checkBlockchainAsync() {
+        await this._checkUserAccountAsync();
+        await this._checkNetworkIdAsync();
+    }
+    private async _checkBalancesAsync() {
+        await this._checkUserBalanceAsync();
+        await this._checkTokenBalanceAsync();
+    }
+
+    private async _checkTokenBalanceAsync() {
+        const currentToken = this._store.getState().selectedToken;
+        const currentUser = this._store.getState().userAddress;
+        const userTokenBalances = this._store.getState().userTokenBalances[currentUser];
+        let tokenBalance = new BigNumber(0);
+
+        // TODO this won't work for the many tokens and NFTs not in the registry
+        const tokenAddress = await this._zeroEx.tokenRegistry.getTokenAddressBySymbolIfExistsAsync(currentToken);
```

Comment:
This is also super inefficient. This should be fetched once and cached. We don't expect to be updating the tokenRegistry frequently and the client shouldn't expect real-time updates... In the off chance this does update, they should just re-fresh the page.

Reply:
Thinking more long term though, really the list of desired tokens tradeable on the Trade Widget should be passed in at component instantiation. Different sites will want to cater to different audiences...